### ISSUE TITLE
Publish to Maven

### ISSUE THREAD

**Issue by cosmincatalin** [awslabs/kinesis-aggregation#10]
Hi,

It would be great if the Java classes will also be made available on Maven.
I have this project [https://github.com/audienceproject/kinesis-writer](https://github.com/audienceproject/kinesis-writer) which I have made available in Maven Central that depends on this.

Best regards,
Cosmin

**Comment by brentnash** [comment]
We originally made a conscious decision not to publish to Maven, but I don't recall what the specific reason was.  Let me check into what would be involved and get back to you.  Thanks for the feedback!

**Comment by tzvetkovg** [comment]
Is there any progress on that? My project relies on the jar and I have to locally upload it rather than use maven..

**Comment by mransley** [comment]
Bit surprised this isn't available as most of the code looks pretty generic.  That said, it has been 11 months!

**Comment by vinaysen** [comment]
We have a strong usecase where we are would need aggregation on Kinesis records and this library fits in perfectly for our scenario. The only issue is that it is not  published on Maven Central.
Are you planning to publish this on Maven? Or, do you recommend any other approach for aggregation/de-aggregation?

**Comment by cosmincatalin** [comment]
One alternative would be for one of us to publish t to Maven.

**Comment by brentnash** [comment]
Do any of you have experience publishing a package to Maven?

I'd love to make that happen, but there was something previously that prevented us from doing so and I don't recall what it was.  It could be that it's no longer an issue.

**Comment by brentnash** [comment]
Quick update:  I have a path forward for this.  I'll update more when the submission process is moving forward.

**Comment by behrangsa** [comment]
Hi @brentnash, any updates?

**Comment by brentnash** [comment]
I just submitted an access request to get the proper publishing credentials.  Waiting for a response.

**Comment by brentnash** [comment]
Access got approved.  Looking into the actual packaging and publishing steps now.

**Comment by ssesha** [comment]
Hi @brentnash, any updates on this?

**Comment by behrangsa** [comment]
@IanMeyers @brentnash any updates on this?

**Comment by ryan-persaud** [comment]
Bump.  Also, I just cloned the repo locally and tried to build master, but there were multiple errors.  This is a decidedly sub-optimal situation.  I just needed the RecordDeaggregator class, so I dropped it into my project directly and added the latest amazon-kinesis-client as a dependency in my pom, instead of trying to fight with local builds.  Hopefully, this can get added to Maven Central soon as it is the officially prescribed method for deaggregating Kinesis records in Lambda functions (http://docs.aws.amazon.com/streams/latest/dev/kinesis-record-deaggregation.html)

**Comment by brentnash** [comment]
That's fair @ryan-persaud .  I was actually not aware that this was being linked from the official documentation.  This started off as a temporary solution side project a couple years ago.

Anyway, to that end, I will see if I can patch up the existing build issues and then I'm dedicating today to the Maven publish.  Maven is unfortunately a more complicated beast than NPM or Pypi and I'm trying to make sure I go through the proper channels.  I should have everything I need now.

**Comment by brentnash** [comment]
@ryan-persaud 

Compile errors should be fixed now
(ref: https://github.com/awslabs/kinesis-aggregation/pull/40)

**Comment by ryan-persaud** [comment]
@brentnash Thanks for getting the ball unstuck on this.

**Comment by brentnash** [comment]
Still rolling along.  Made some fixes (see commit history) and have the javadoc/source aggregates working.  Trying to do the same to make the compiled aggregate jar at the moment.  Then will try uploading to sonatype staging.

**Comment by brentnash** [comment]
The current Maven structure of this project that mixes the root pom.xml into both a parent project and an aggregator project is making Maven uploads incredibly tedious.  I'm in the process of condensing it down to just an aggregator.  

I also believe that only the KinesisAggregator and KinesisDeaggregator should actually be going to Maven central; the result of this GitHub project is to provide sample code that's supposed to be a reference rather than distributable packages.

The current plan, in order to unblock everyone as quickly as possible, will be to upload the KinesisAggregator and KinesisDeaggregator as two separate projects on Maven.

Current progress is happening on this branch:  https://github.com/awslabs/kinesis-aggregation/tree/maven-reorg

**Comment by ryan-persaud** [comment]
@brentnash Thanks for the update - let me know if I can help in any way.

**Comment by brentnash** [comment]
I think I have all the Maven tweaks necessary in the maven-reorg branch now.  My next step is going to be uploading to the Maven staging environment per http://central.sonatype.org/pages/ossrh-guide.html#deployment

Assuming the staging upload succeeds and is usable, I'll plan on merging the "maven-reorg" branch into master and also pushing the Maven projects live.

**Comment by brentnash** [comment]
OK, good news I think.  I've got the packages published to the Maven **staging** repository and was able to throw together this POM:

```
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
	<modelVersion>4.0.0</modelVersion>
	<groupId>com.amazonaws.foo</groupId>
	<artifactId>agg-test</artifactId>
	<version>0.0.1-SNAPSHOT</version>

	<repositories>
		<repository>
			<id>sonatype-nexus-staging-agg</id>
			<name>Sonatype Staging Repository</name>
			<url>REDACTED</url>
		</repository>
		<repository>
			<id>sonatype-nexus-staging-deagg</id>
			<name>Sonatype Staging Repository</name>
			<url>REDACTED</url>
		</repository>
	</repositories>

	<dependencies>
		<dependency>
			<groupId>com.amazonaws.services.kinesis</groupId>
			<artifactId>kinesis-agg</artifactId>
			<version>1.0.3</version>
		</dependency>
		<dependency>
			<groupId>com.amazonaws.services.kinesis</groupId>
			<artifactId>kinesis-deagg</artifactId>
			<version>1.0.3</version>
		</dependency>
	</dependencies>

</project>
```

And with that POM pulling from staging, I threw together this quick proof-of-concept that seems to work (imports left out for brevity):

```
public class App
{
	public static void main(final String[] args) throws Exception
	{
		//create dummy records
		String inputRecord1 = "abcdefghijklmnopqrstuvwxyz";
		System.out.println(inputRecord1);
		String inputRecord2 = "zyxwvutsrqponmlkjihgfedcba";
		System.out.println(inputRecord2);
		
		//aggregate into a single record
		RecordAggregator ra = new RecordAggregator();
		ra.addUserRecord("pk", inputRecord1.getBytes(StandardCharsets.UTF_8));
		ra.addUserRecord("pk", inputRecord2.getBytes(StandardCharsets.UTF_8));
		AggRecord aggRecord = ra.clearAndGet();
		byte[] recordBytes = aggRecord.toRecordBytes();
		
		//create a fake Lambda Kinesis event record
		KinesisEventRecord ker = new KinesisEventRecord();
		KinesisEvent.Record r = new KinesisEvent.Record();
		r.setData(ByteBuffer.wrap(recordBytes));
		ker.setKinesis(r);
		
		//deaggregate records
		List<UserRecord> userRecords = RecordDeaggregator.deaggregate(ker);
		for(UserRecord userRecord : userRecords)
		{
			byte[] outBytes = userRecord.getData().array();
			System.out.println(new String(outBytes, StandardCharsets.UTF_8));
		}
	}
}
```

Based on this I'm going to merge the Maven reorg branch into master and also publish these live on Maven.  Stay tuned.

**Comment by brentnash** [comment]
Missed a step.  Filed tickets and currently waiting on SonaType to establish repositories:

https://issues.sonatype.org/browse/OSSRH-36708
https://issues.sonatype.org/browse/OSSRH-36709

Not 100% sure if this is required or not (so if anyone on this issue knows otherwise, let me know).  Otherwise they're committed to a 2-day SLA.

**Comment by brentnash** [comment]
Tickets are resolved.  I published from staging.  Waiting for the artifacts to appear in central.

**Comment by brentnash** [comment]
Looks like the sync to Maven Central has completed.  There are two projects:

http://central.maven.org/maven2/com/amazonaws/amazon-kinesis-aggregator/
https://search.maven.org/#search%7Cga%7C1%7Ca%3A%22amazon-kinesis-aggregator%22

http://central.maven.org/maven2/com/amazonaws/amazon-kinesis-deaggregator/
https://search.maven.org/#search%7Cga%7C1%7Ca%3A%22amazon-kinesis-deaggregator%22

I ran a test project with them included that seems to be working.  You can try the same on your end by adding these dependencies to your pom.xml:

```
<dependencies>
		<dependency>
			<groupId>com.amazonaws</groupId>
			<artifactId>amazon-kinesis-aggregator</artifactId>
			<version>1.0.3</version>
		</dependency>
		<dependency>
			<groupId>com.amazonaws</groupId>
			<artifactId>amazon-kinesis-deaggregator</artifactId>
			<version>1.0.3</version>
		</dependency>
	</dependencies>
```

I'm going to merge the maven-reorg branch back into master now.  When all that is complete I'll close the issue.

**Comment by brentnash** [comment]
Pull request merged into Master:

https://github.com/awslabs/kinesis-aggregation/pull/41/files

**Comment by brentnash** [comment]
Fresh clone of master seems to build properly.  I think we can officially close this one out.  Feel free to re-open if/when you discover issues.
